src/main/flight/imu.c

```diff
@@ -215,51 +213,79 @@ static float invSqrt(float x)
+    if (gpsGroundspeedGain > 0.0f) {
+        // Compute GPS heading vector in earth frame (ef) from scalar GPS courseOverGround
+        // Earth frame X is pointing north and sin/cos argument is anticlockwise
         const fpVector2_t cog_ef = {.x = cos_approx(-courseOverGround), .y = sin_approx(-courseOverGround)};
-#define THRUST_COG 1
-#if THRUST_COG
-        const fpVector2_t heading_ef = {.x = rMat[X][Z], .y = rMat[Y][Z]};  // body Z axis (up) - direction of thrust vector
-#else
-        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};  // body X axis. Projected vector magnitude is reduced as pitch increases
-#endif
-        // cross product = 1 * |heading| * sin(angle) (magnitude of Z vector in 3D)
-        // order operands so that rotation is in direction of zero error
+
+        // Compute IMU / craft heading vector and magnitude in body X axis
+        // Important:  vector magnitude is reduced as pitch increases
+        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};
+        const float headingMagnitude = vector2Mag(&heading_ef);
+
+        // cross (product) = heading * cogHeading * sin error angle, ie, reflects error angle of quad X axis vs GPS groundspeed
+        // cross value sign depends on the rotation change from input to output vectors by right hand rule
+        // operand order is such that rotation is in direction of zero error
+        // abs value of cross is zero when parallel, max of 1.0 at 90 degree error
+        // cross value determines ez_ef when error is less than 90 degrees
+        // a zero cross value at 180 degree error would prevent error correction
         const float cross = vector2Cross(&heading_ef, &cog_ef);
-        // dot product, 1 * |heading| * cos(angle)
+
+        // dot product = heading * cogHeading * cos(angle)
+        // max when aligned, zero at 90 degrees of vector difference, negative for any error > 90 degrees
         const float dot = vector2Dot(&heading_ef, &cog_ef);
-        // use cross product / sin(angle) when error < 90deg (cos > 0),
-        //   |heading| if error is larger (cos < 0)
-        const float heading_mag = vector2Mag(&heading_ef);
-        float ez_ef = (dot > 0) ? cross : (cross < 0 ? -1.0f : 1.0f) * heading_mag;
-#if THRUST_COG
-        // increase gain for small tilt (just heuristic; sqrt is cheap on F4+)
-        ez_ef /= sqrtf(heading_mag);
-#endif
-        ez_ef *= cogYawGain;          // apply gain parameter
+
+
+        // ez_ef is the error value used to correct the heading via an IIR like smoothing function
+        // for error angles less than 90 degrees (dot > 0), use cross value to compute ez_ef
+        // above 90 degrees error, use heading vector magnitude, but give it sign from cross product
+        // heading vector magnitude is 1.0 when flat, and is attenuated when pitch angle is large
+        // when error exceeds 90 degrees, and pitch angle is high, a transient step may occur in ez_ef
```

Comment:
There should be no transient (see above). Is is a bug otherwise

Reply:
At 90deg it changes from sin(90) to sign(90) -> 1 to 1. No discontinuity here

Reply:
Even when the quad is, say, pitched forward at 45 degrees when we make this transition?
You previously made a note that `headingMagnitude` is reduced towards zero in the presence of a non-zero pitch angle.  Wouldn't `headingMagnitude` be around 0.7 when pitch angle is 45 degrees?  Meaning an abrupt transition from 1.0 to 0.7?
Maybe the note:
`// Important:  vector magnitude is reduced as pitch increases`
Does not apply because we are now calculating heading `in body X axis`?
`// Compute IMU / craft heading vector and magnitude in body X axis`

Reply:
cross product already has `headingMagnitude` factored in, it will be around .7 at 90 deg ( [1 0] cross [0 0.7] == 0.7 )

Reply:
This is very confusing.  How can we use the cross value as an effective measure of the angular error of the heading, if it changes with pitch?  The current code sets ez_ef to cross.  But if cross changes just with a pitch input, and nothing else, and we know that a pure pitch change does not change heading - how can that work?

Reply:
But ez_ef is _error_ estimate. Normally, it is zero and pitch has no influence. 
It is similar to changing P factor of PID - when you halve it, error of smaller, but PID still works.